### PR TITLE
Hides nodes from listing that the user has no access to

### LINKED ISSUE #22578: Calendar names can be enumerated
- Request a not existing calendar: `http://localhost/remote.php/dav/calendars/admin/NotExistingCalendar`

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\NotFound</s:exception>
  <s:message>Node with name 'NotExistingCalendar' could not be found</s:message>
</d:error>
```

- Request an existing, not-shared, calendar:  `http://localhost/remote.php/dav/calendars/admin/existingcalendar`
```xml

<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAVACL\Exception\NeedPrivileges</s:exception>
  <s:message>User did not have the required privileges ({DAV:}read) for path "calendars/admin/existingcalendar"</s:message>
  <d:need-privileges>
    <d:resource>
      <d:href>/remote.php/dav/calendars/admin/existingcalendar</d:href>
      <d:privilege>
        <d:read/>
      </d:privilege>
    </d:resource>
  </d:need-privileges>
</d:error>
```

(probably also affects the other endpoints, just reading through the code and testing)

<hr/>

From a security PoV we should not leak the existence of a calendar. Is there any CalDAV standard that prohibits this? It seems odd to me to leak the existence of such things.

cc @DeepDiver1975

### PR DESCRIPTION
**Pull request by LukasReschke**
Fixes  https://github.com/owncloud/core/issues/22578, most of it are integration tests. The integration tests are not nice but do work :tm: 

… we need at some point clean them up a bit, the existing structure with the "use" trait is rather strange IMHO …

@DeepDiver1975 THX

### CODE REVIEW

**Comment by LukasReschke** [comment]
Ok. I'll adjust this to completely cover #22578. Should be easier than thought…

**Comment by LukasReschke** [comment]
@DeepDiver1975 Please take a look. THX.

**Comment by rullzer** [comment]
Nice. Tested a few webdav calls and 404 is returned properly now. (So Basically did the intergration tests by hand :P).

:+1: 

And yes we need to look at the intergration tests ;)

**Comment by LukasReschke** [comment]
Yay. The integration tests are failing due to the fact that https://github.com/php/php-src/blob/97294aca7e066443291cc2d77f8674ac23eabb32/sapi/cli/php_http_parser.c#L78-L105 does not support MKCALENDAR. Adjusting this PR and filing one to PHP then……

**Comment by LukasReschke** [comment]
Well. There is already one at https://github.com/php/php-src/pull/1417

**Comment by blizzz** [comment]
Tested using cadaver, works :+1:

File: apps/dav/lib/connector/sabre/davaclplugin.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	function checkPrivileges($uri, $privileges, $recursion = self::R_PARENT, $throwExceptions = true) {
+		$access = parent::checkPrivileges($uri, $privileges, $recursion, false);
+		if($access === false) {
+			/** @var INode $node */
+			$node = $this->server->tree->getNodeForPath($uri);
+
+			switch(get_class($node)) {
+				case 'OCA\DAV\CardDAV\AddressBook':
```
Comment (DeepDiver1975) [inline apps/dav/lib/connector/sabre/davaclplugin.php]: what about calendars?

Comment (LukasReschke) [inline apps/dav/lib/connector/sabre/davaclplugin.php]: SabreDAV uses "Node" there as error message which is the default. Apparently only for AddressBooks it uses a custom error message.

We have integration tests in place below to ensure that we notice the case when SabreDAV changes the error message.

**Pull request merged**